client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -259,12 +260,23 @@ public void flush() {
                         // Segment sealed exception observed during a flush. Re-run flush on all the
                         // available writers.
                         success = false;
-                        log.warn("Flush on segment {} failed due to {}, it will be retried.", writer.getSegmentName(), e.getMessage());
+                        log.warn("Flush on segment {} by event writer {} failed due to {}, it will be retried.",
+                                 writer.getSegmentName(), writerId, e.getMessage());
                         tryWaitForSuccessors();
                         break;
+                    } catch (RetriesExhaustedException e1) {
+                        // Ensure a flush is invoked on all the segment writers before throwing a RetriesExhaustedException.
+                        log.warn("Flush on segment {} by event writer {} failed after all configured retries",
+                                 writer.getSegmentName(), writerId);
+                        retriesExhaustedException = e1;
                     }
                 }
             }
+            if (retriesExhaustedException != null) {
+                log.error("Flush by writer {} on Stream {} failed after all retries to connect with Pravega exhausted.",
+                          writerId, stream.getScopedName());
+                throw retriesExhaustedException;
+            }
```

Comment:
This logic strikes me as wrong. It catches retriesExhausted and then doesn't break out of the loop so it keeps retrying until it succeeds and then it throws retriesExhausted anyway even though it was successfully written. 
I don't see the point in catching and rethrowing. If the exception is going to be propagated upward, there is no need to re-log it at every stack frame along the way, just let it go up.

Reply:
The reason for this change is to handle a scenario where out of the many segment writers only a subset of them face a connectivity issue with a specific segment store instance. This logic would ensure we ensure flush is invoked on every segment writer before throwing a RetriesExhaustedException to the application. 
(This would ensure events from segment writers who still have connectivity with SegmentStores can flush all their inflight events before the application is intimated about the connectivity

Reply:
ok. But it doesn't do that. It will keep retrying as long as any segment fails to write forever. And it may throw an error if it succeeds.